src/digraph-cyclicity.jl

```diff
@@ -0,0 +1,320 @@
+abstract type Visitor{T<:Integer} end
+
+"""
+    ncycles_n_i(n::Integer, i::Integer)
+
+Compute the theoretical maximum number of cycles of size `i` in a directed graph of `n`
+ vertices.
+"""
+ncycles_n_i(n::Integer, i::Integer) = binomial(big(n), big(n-i+1)) * factorial(big(n-i))
+
+
+"""
+    maxsimplecycles(n::Integer)
+
+Compute the theoretical maximum number of cycles in a directed graph of `n` vertices,
+assuming there are no self-loops.
+The formula is coming from [Johnson's paper](http://epubs.siam.org/doi/abs/10.1137/0204007).
+"""
+
+maxsimplecycles(n::Integer) = sum(x -> ncycles_n_i(n, x), 1:(n-1))
+
+
+@doc_str """
+    maxsimplecycles(dg::DiGraph, byscc::Bool = true)
+
+Compute the theoretical maximum number of cycles in the directed graph `dg`.
+
+The computation can be performed assuming the graph is complete or taking into account the
+decomposition in strongly connected components (`byscc` parameter). The formula is coming from
+[Johnson's paper](http://epubs.siam.org/doi/abs/10.1137/0204007).
+
+### Performance
+A more efficient version is possible.
+"""
+function maxsimplecycles end
+@traitfn function maxsimplecycles(dg::::IsDirected, byscc::Bool = true)
+    c = 0
+    n = nv(dg)
+    if !byscc
+        c = maxsimplecycles(n)
+    else
+        for scc in strongly_connected_components(dg)
+            if length(scc) > 1
+                c += maxsimplecycles(length(scc))
+            end
+        end
+    end
+    return c
+end
+
+"""
+```
+type JohnsonVisitor <: Visitor
+stack::Vector{Int}
+blocked::Vector{Bool}
+blockedmap::Vector{Set{Int}}
+end
+```
+
+Composite type that regroups the information needed for Johnson's algorithm.
+
+`stack` is the stack of visited vertices. `blocked` is a boolean for each 
+vertex that tells whether it is blocked or not. `blockedmap` tells which 
+vertices to unblock if the key vertex is unblocked.
+"""
+type JohnsonVisitor{T<:Integer} <: Visitor{T}
+    stack::Vector{T}
+    blocked::BitArray
+    blockedmap::Vector{Set{T}}
+end
+
+"""
+    JohnsonVisitor(dg::::IsDirected)
+
+Constructor of the visitor, using the directed graph information.
+"""
+JohnsonVisitor{T<:Integer}(dg::DiGraph{T}) = JohnsonVisitor(Vector{T}(),

```

Comment:
I have not been able to it. I will try to make clear what my problem is:
I declare the new type JohnsonVisitor with a type T. I need that type in the definition of the constructor. If I use trait IsDirected, I don't know how to get the type out of it.
Am I clear?